File: dataplane/internal/core/budget.go
package core

// BudgetID names a spend limit defined in the control plane.
type BudgetID string

// BudgetScope says which level of the identity hierarchy a budget attaches to.
//
// Training is its own scope, not a line item under inference. A single
// fine-tuning run can cost more than a month of serving, so it gets its own
// limit and its own approval threshold.
type BudgetScope uint8

// The scopes a budget may attach to. Unset is the zero value and is never
// valid in a built layer.
const (
	BudgetScopeUnset BudgetScope = iota
	BudgetScopeKey
	BudgetScopeApp
	BudgetScopeUser
	BudgetScopeTeam
	BudgetScopeOrg
	BudgetScopeModel
	BudgetScopeTraining
)

// BudgetRef is one link in a principal's budget chain. The chain is precomputed
// into the Principal so that admission is an array scan over a handful of
// integers rather than a graph walk.
//
// A request must satisfy *every* budget in its chain.
type BudgetRef struct {
	ID    BudgetID
	Scope BudgetScope
}

// BudgetState is a budget's limit and its spend as of the snapshot that carries
// it. Budgets are eventually consistent by design: usage events flow to an
// accounting consumer, which folds the result into the next snapshot. Rate
// limits are the mechanism for anything that must be immediate.
type BudgetState struct {
	ID    BudgetID
	Scope BudgetScope

	LimitMicroUSD MicroUSD
	SpentMicroUSD MicroUSD

	// Hard budgets deny on exhaustion. Soft budgets emit a warning header and an
	// event, and let the request through.
	Hard bool

	// HeadroomBasisPoints holds back a slice of a hard limit so that requests
	// already streaming when the budget tips can finish without overshooting.
	// 500 = 5%, which is the documented default. uint32 to match the wire
	// format, for the same reason Deployment.Weight is.
	HeadroomBasisPoints uint32
}

// Available is the spend remaining before the headroom band begins.
func (b BudgetState) Available() MicroUSD {
	reserved := b.LimitMicroUSD * MicroUSD(b.HeadroomBasisPoints) / 10_000
	remaining := b.LimitMicroUSD - reserved - b.SpentMicroUSD
	if remaining < 0 {
		return 0
	}
	return remaining
}

// Denies reports whether this budget must reject a new request. Soft budgets
// never deny; they are observed at admission and reported, not enforced.
func (b BudgetState) Denies() bool { return b.Hard && b.Available() == 0 }

File: controlplane/src/model_gateway_control/domain/budget.py
"""Spend limits."""

from __future__ import annotations

from dataclasses import dataclass
from enum import IntEnum

from model_gateway_control.errors import InvalidRequestError

#: Fraction of a hard limit held back, in basis points, so that requests
#: already streaming when a budget tips can finish without overshooting.
DEFAULT_HEADROOM_BASIS_POINTS = 500


class BudgetScope(IntEnum):
    """Which level of the identity hierarchy a budget attaches to.

    ``TRAINING`` is its own scope rather than a line item under inference: a
    single fine-tuning run can cost more than a month of serving, so it gets its
    own limit and its own approval threshold.
    """

    UNSET = 0
    KEY = 1
    APP = 2
    USER = 3
    TEAM = 4
    ORG = 5
    MODEL = 6
    TRAINING = 7


@dataclass(frozen=True, slots=True, kw_only=True)
class Budget:
    """A limit and what has been spent against it.

    Budgets are eventually consistent by design: usage events flow to an
    accounting consumer, which folds the result into the next snapshot. Rate
    limits are the mechanism for anything that must be immediate.
    """

    id: str
    scope: BudgetScope
    limit_micro_usd: int
    spent_micro_usd: int = 0
    #: Hard budgets deny on exhaustion; soft ones warn and let the request pass.
    hard: bool = True
    headroom_basis_points: int = DEFAULT_HEADROOM_BASIS_POINTS

    def __post_init__(self) -> None:
        if not self.id:
            raise InvalidRequestError("a budget needs an id")
        if self.scope is BudgetScope.UNSET:
            raise InvalidRequestError(f"budget {self.id!r} has an unset scope")
        if self.limit_micro_usd < 0 or self.spent_micro_usd < 0:
            raise InvalidRequestError(f"budget {self.id!r} has a negative amount")
        if not 0 <= self.headroom_basis_points <= 10_000:
            raise InvalidRequestError(f"budget {self.id!r} headroom is out of range")

    @property
    def available_micro_usd(self) -> int:
        """Spend remaining before the headroom band begins."""
        reserved = self.limit_micro_usd * self.headroom_basis_points // 10_000
        return max(0, self.limit_micro_usd - reserved - self.spent_micro_usd)
